README.md
# consul_kv_dart

Dart client for the Consul KV-store.

Wraps Consul's simple [key/value](https://www.consul.io/api/kv.html) store API.

Client uses HTTP requests to communicate with Consul.

Each method has optional parameters that are described in Consul API docs.

You might want to wrap responses into some kind of handler 
(check status codes, encode\convert values, etc.) - result is a flexible [`Response`](https://www.dartdocs.org/documentation/http/0.11.3%2B16/http/Response-class.html) object.

Each method return [`Future`](https://api.dartlang.org/stable/1.24.3/dart-async/Future-class.html) which is used to get result.

Additional documentation can be found [here](https://www.dartdocs.org/documentation/consul_kv_dart/1.0.2/).

## Usage

A simple usage example:

```dart
Future main() async {
  var consul = new ConsulKV(port: 5579, defaultHeaders: {'X-Consul-Token': 'bcd1234'});
  var newKey = await consul.put('foo', 'bar');
  var foundKey = await consul.get('foo');
  var insertIfNotExists = await consul.put('foo', 'bar', cas: 1);
  var deleteKey = await consul.delete('foo');
  var keyNotExists = await consul.get('foo');

  print(newKey.body); // true
  print(newKey.statusCode); // 200

  print(foundKey.body); // [{"LockIndex":0,"Key":"foo","Value":"YmFy",
                        // "CreateIndex":208877,"ModifyIndex":208877}]
  print(foundKey.statusCode);  // 200

  print(insertIfNotExists.body); // false
  print(insertIfNotExists.statusCode); // 200

  print(deleteKey.body); // true
  print(deleteKey.statusCode); // 200

  print(keyNotExists.statusCode); // 404

  consul.close();

}
```



CHANGELOG.md
# Changelog

## 1.0.2

- Resolving docgen issues

## 1.0.1

- Description update

## 1.0.0

- Initial version


test/consul_kv_dart_test.dart
import 'package:consul_kv_dart/consul_kv_dart.dart';
import 'package:http/http.dart';
import 'package:http/testing.dart';
import 'package:test/test.dart';

void main() {
  group('Consul KV client', () {
    ConsulKV client;
    MockClient mockClient;
    Map<String, String> respHeaders;
    var getResponse =
        '[{"LockIndex":0,"Key":"dpg/user/display_limits/bar_chart","Flags":0,"Value":"MTAwMA==","CreateIndex":191430,"ModifyIndex":191430}]';

    setUp(() {
      respHeaders = {
        "content-type": "application/json",
        "date": "Wed, 17 Jan 2018 10:40:48 GMT",
        "x-consul-knownleader": "true",
        "x-consul-index": "191430",
        "x-consul-lastcontact": "0"
      };
      client = new ConsulKV(
          host: '127.0.0.1', defaultHeaders: {'X-Consul-Token': 'bcd1234'});
      mockClient = new MockClient((request) async {
        if (request.method == 'GET' && request.url.path == '/v1/kv/testkey') {
          return new Response(getResponse, 200, headers: respHeaders);
        }
        ;
        if (request.method == 'GET' &&
            request.url.path != '/v1/kv/testkey' &&
            request.url.path != '/v1/kv/') {
          return new Response('', 401, headers: respHeaders);
        }
        ;
        if (request.method == 'GET' && request.url.path == '/v1/kv/') {
          return new Response('Missing key name', 400, headers: respHeaders);
        }
        ;
        if (request.method == 'PUT' &&
            request.url.path == '/v1/kv/testkey' &&
            !request.url.queryParameters.containsKey('cas')) {
          return new Response('true', 200, headers: respHeaders);
        }
        ;
        if (request.method == 'PUT' &&
            request.url.path == '/v1/kv/testkey' &&
            request.url.queryParameters['cas'] == '1') {
          return new Response('false', 200, headers: respHeaders);
        }
        ;
        if (request.method == 'DELETE' &&
            request.url.path == '/v1/kv/testkey' &&
            !request.url.queryParameters.containsKey('cas')) {
          return new Response('true', 200, headers: respHeaders);
        }
        ;
        if (request.method == 'DELETE' &&
            request.url.path == '/v1/kv/testkey' &&
            request.url.queryParameters['cas'] == '1') {
          return new Response('false', 200, headers: respHeaders);
        }
        ;
        return new Response("", 400, headers: respHeaders);
      });
      client.switchClient = mockClient;
    });
    test('Checks that optional params are present in GET method', () async {
      var resultOne = await client
          .get('testkey', headers: {"content-type": "application/json"});
      var resultTwo = await client.get('testkey', recurse: true);
      var resultThree = await client.get('testkey', keys: true);
      var resultFour = await client.get('testkey', dc: 'dc1');
      var resultFive = await client.get('testkey', separator: ':');
      var resultSix = await client.get('testkey',
          separator: ':', recurse: true, keys: true, dc: 'dc1');
      expect(resultOne.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?separator=%2F'));
      expect(resultOne.request.url.queryParameters, equals({'separator': '/'}));
      expect(
          resultOne.request.headers,
          equals({
            'content-type': 'application/json',
            'X-Consul-Token': 'bcd1234'
          }));

      expect(resultTwo.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?recurse&separator=%2F'));
      expect(resultTwo.request.url.queryParameters,
          equals({'separator': '/', 'recurse': ''}));

      expect(resultThree.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?keys&separator=%2F'));
      expect(resultThree.request.url.queryParameters,
          equals({'separator': '/', 'keys': ''}));

      expect(resultFour.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?dc=dc1&separator=%2F'));
      expect(resultFour.request.url.queryParameters,
          equals({'separator': '/', 'dc': 'dc1'}));

      expect(resultFive.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?separator=%3A'));
      expect(
          resultFive.request.url.queryParameters, equals({'separator': ':'}));

      expect(
          resultSix.request.url.toString(),
          equals(
              'http://127.0.0.1:8500/v1/kv/testkey?dc=dc1&recurse&keys&separator=%3A'));
      expect(resultSix.request.url.queryParameters,
          equals({'dc': 'dc1', 'recurse': '', 'keys': '', 'separator': ':'}));
    });
    test('Attempts to get key from Consul with different params', () async {
      var result = await client.get('testkey');
      expect(result.request.url.path, equals('/v1/kv/testkey'));
      expect(result.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?separator=%2F'));
      expect(result.request.headers, equals({'X-Consul-Token': 'bcd1234'}));
      expect(result.headers, equals(respHeaders));
      expect(result.body, equals(getResponse));
    });
    test('Attempts to get not existing key returns proper response', () async {
      var result = await client.get('notexistingkey');
      expect(result.request.url.path, equals('/v1/kv/notexistingkey'));
      expect(result.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/notexistingkey?separator=%2F'));
      expect(result.request.headers, equals({'X-Consul-Token': 'bcd1234'}));
      expect(result.headers, equals(respHeaders));
      expect(result.statusCode, equals(401));
    });
    test('Attempts to get key with no params returns proper response',
        () async {
      var result = await client.get('');
      expect(result.request.url.path, equals('/v1/kv/'));
      expect(result.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/?separator=%2F'));
      expect(result.request.headers, equals({'X-Consul-Token': 'bcd1234'}));
      expect(result.headers, equals(respHeaders));
      expect(result.body, equals('Missing key name'));
      expect(result.statusCode, equals(400));
    });
    test('Checks that optional params are present in PUT method', () async {
      var resultOne = await client.put('testkey', 'testvalue');
      var resultTwo = await client.put('testkey', 'testvalue',
          headers: {'content-type': 'application/json'});
      var resultThree = await client.put('testkey', 'testvalue', dc: 'dc1');
      var resultFour = await client.put('testkey', 'testvalue', flags: 1);
      var resultFive = await client.put('testkey', 'testvalue', cas: 1);
      var resultSix = await client.put('testkey', 'testvalue', acquire: "test");
      var resultSeven =
          await client.put('testkey', 'testvalue', release: "test");

      expect(resultOne.request.url.path, equals('/v1/kv/testkey'));
      expect(resultOne.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?'));
      expect(
          resultOne.request.headers,
          equals({
            'X-Consul-Token': 'bcd1234',
            'content-type': 'text/plain; charset=utf-8'
          }));

      expect(
          resultTwo.request.headers,
          equals({
            'content-type': 'application/json; charset=utf-8',
            'X-Consul-Token': 'bcd1234'
          }));

      expect(resultThree.request.url.queryParameters, equals({'dc': 'dc1'}));

      expect(resultFour.request.url.queryParameters, equals({'flags': '1'}));

      expect(resultFive.request.url.queryParameters, equals({'cas': '1'}));

      expect(
          resultSix.request.url.queryParameters, equals({'acquire': 'test'}));

      expect(
          resultSeven.request.url.queryParameters, equals({'release': 'test'}));
    });
    test('Attempts to put key returns proper response', () async {
      var resultOne = await client.put('testkey', 'testvalue');
      var resultTwo = await client.put('testkey', 'testvalue', cas: 1);
      expect(resultOne.request.url.path, equals('/v1/kv/testkey'));
      expect(resultOne.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?'));
      expect(resultOne.statusCode, equals(200));
      expect(
          resultOne.request.headers,
          equals({
            'X-Consul-Token': 'bcd1234',
            'content-type': 'text/plain; charset=utf-8'
          }));
      expect(resultOne.body, equals('true'));
      expect(resultOne.headers, equals(respHeaders));

      expect(resultTwo.request.url.path, equals('/v1/kv/testkey'));
      expect(resultTwo.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?cas=1'));
      expect(resultTwo.statusCode, equals(200));
      expect(
          resultTwo.request.headers,
          equals({
            'X-Consul-Token': 'bcd1234',
            'content-type': 'text/plain; charset=utf-8'
          }));
      expect(resultTwo.body, equals('false'));
      expect(resultTwo.headers, equals(respHeaders));
      expect(resultTwo.request.url.queryParameters, equals({'cas': '1'}));
    });
    test('Checks that optional params are present in DELETE method', () async {
      var resultOne = await client.delete('testkey');
      var resultTwo = await client
          .delete('testkey', headers: {'content-type': 'application/json'});
      var resultThree = await client.delete('testkey', recurse: true);
      var resultFour = await client.delete('testkey', cas: 1);

      expect(resultOne.request.url.path, equals('/v1/kv/testkey'));
      expect(resultOne.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?'));
      expect(resultOne.request.headers, equals({'X-Consul-Token': 'bcd1234'}));

      expect(
          resultTwo.request.headers,
          equals({
            'content-type': 'application/json',
            'X-Consul-Token': 'bcd1234'
          }));

      expect(resultThree.request.url.queryParameters, equals({'recurse': ''}));

      expect(resultFour.request.url.queryParameters, equals({'cas': '1'}));
    });
    test('Attempts to delete key returns proper response', () async {
      var resultOne = await client.delete('testkey');
      var resultTwo = await client.put('testkey', 'testvalue', cas: 1);
      expect(resultOne.request.url.path, equals('/v1/kv/testkey'));
      expect(resultOne.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?'));
      expect(resultOne.statusCode, equals(200));
      expect(resultOne.request.headers, equals({'X-Consul-Token': 'bcd1234'}));
      expect(resultOne.body, equals('true'));
      expect(resultOne.headers, equals(respHeaders));

      expect(resultTwo.request.url.path, equals('/v1/kv/testkey'));
      expect(resultTwo.request.url.toString(),
          equals('http://127.0.0.1:8500/v1/kv/testkey?cas=1'));
      expect(resultTwo.statusCode, equals(200));
      expect(
          resultTwo.request.headers,
          equals({
            'X-Consul-Token': 'bcd1234',
            'content-type': 'text/plain; charset=utf-8'
          }));
      expect(resultTwo.body, equals('false'));
      expect(resultTwo.headers, equals(respHeaders));
      expect(resultTwo.request.url.queryParameters, equals({'cas': '1'}));
    });
    test('Checks that getClient() getter returns current HTTP Client',
        () async {
      var newClient = new Client();
      var currentClient = client.currentClient;
      expect((currentClient.hashCode == newClient.hashCode), equals(false));
      expect(currentClient.hashCode, equals(mockClient.hashCode));
    });
    test(
        'Checks that switchClient() setter replaces current HTTP Client with new one',
        () async {
      var newClient = new Client();
      var currentClient = client.currentClient;
      expect((currentClient.hashCode == newClient.hashCode), equals(false));
      expect(currentClient.hashCode, equals(mockClient.hashCode));
      client.switchClient = newClient;
      currentClient = client.currentClient;
      expect(currentClient.hashCode, equals(newClient.hashCode));
    });
  });
}


example/consul_kv_dart_example.dart
/// Simple example of usage

import 'dart:async';
import 'package:consul_kv_dart/consul_kv_dart.dart';

Future main() async {
  var consul = new ConsulKV(defaultHeaders: {'X-Consul-Token': 'bcd1234'});
  var newKey = await consul.put('foo', 'bar');
  var foundKey = await consul.get('foo');
  var insertIfNotExists = await consul.put('foo', 'bar', cas: 1);
  var deleteKey = await consul.delete('foo');
  var keyNotExists = await consul.get('foo');

  print(newKey.body); // true
  print(newKey.statusCode); // 200

  print(foundKey.body); // [{"LockIndex":0,"Key":"foo","Value":"YmFy",
                        // "CreateIndex":208877,"ModifyIndex":208877}]
  print(foundKey.statusCode);  // 200

  print(insertIfNotExists.body); // false
  print(insertIfNotExists.statusCode); // 200

  print(deleteKey.body); // true
  print(deleteKey.statusCode); // 200

  print(keyNotExists.statusCode); // 404

  consul.close();

}


lib/src/consul_kv_dart_base.dart
/// Implementation of key-value storage client for Consul.
import 'dart:async';

import 'package:http/http.dart' as http;

/// A class that wraps HTTP requests for Consul KV API.
///
/// `http://localhost:8500` is used as default URI when no parameters
/// passed to the constructor.
class ConsulKV {
  /// Host where Consul API is running, defaults to `localhost`.
  String host;

  /// Port where Consul API is running, defaults to `8500`.
  int port;

  /// Scheme for connection, defaults to `http`.
  String scheme;

  /// Consul api version - `v1`.
  String _apiVersion = '/v1';

  /// Endpoint for Consul KV API - `/kv/`.
  String _path = '/kv/';

  /// HTTP Client to perform calls to the API.
  http.Client _client = new http.Client();

  /// headers to use in each request, default to `null`.
  Map<String, String> defaultHeaders;

  /// Creates new instance of Consul KV client.
  ConsulKV(
      {this.host = 'localhost',
      this.port = 8500,
      this.scheme = 'http',
      this.defaultHeaders});

  /// Current [http.Client] of this [ConsulKV].
  http.Client get currentClient => this._client;

  /// New [http.Client] for this [ConsulKV].
  void set switchClient(http.Client client) => this._client = client;

  /// Invokes `close` method from [http.Client] to close connection.
  ///
  ///     var consul = new ConsulKV();
  ///     // do stuff
  ///     consul.close();
  void close() => this._client.close();

  /// Deletes key from KV.
  ///
  /// Parameter [key] is the only required parameter.
  /// Other parameters are optional.
  /// You can read when to use parameters here:
  /// [delete key](https://www.consul.io/api/kv.html#delete-key).
  ///
  ///     var consul = new ConsulKV();
  ///     var deleteKey = await consul.delete('foo');
  ///     print(deleteKey.body); // true
  ///     print(deleteKey.statusCode); // 200
  Future<http.Response> delete(String key,
      {bool recurse, int cas, Map<String, String> headers}) async {
    var params = <String, dynamic>{};
    headers = this._setHeaders(headers);
    if (recurse != null) {
      params['recurse'] = '';
    }
    if (cas != null) {
      params['cas'] = cas.toString();
    }

    var uri = this._buildUri(key, params);
    return (await this._client.delete(uri, headers: headers));
  }

  /// Reads key from KV.
  ///
  /// Parameter [key] is the only required parameter.
  /// Other parameters are optional.
  /// You can read when to use parameters here:
  /// [read key](https://www.consul.io/api/kv.html#read-key).
  ///
  ///     var consul = new ConsulKV();
  ///     var foundKey = await consul.get('foo');
  ///     print(foundKey.body); // [{"LockIndex":0,"Key":"foo","Value":"YmFy",
  ///                             // "CreateIndex":208877,"ModifyIndex":208877}]
  ///     print(foundKey.statusCode);  // 200
  Future<http.Response> get(String key,
      {String dc,
      bool recurse,
      bool raw,
      bool keys,
      String separator = '/',
      Map<String, String> headers}) async {
    var params = <String, String>{};
    headers = this._setHeaders(headers);
    if (dc != null) {
      params['dc'] = dc;
    }
    if (recurse != null) {
      params['recurse'] = '';
    }
    if (raw != null) {
      params['raw'] = '';
    }
    if (keys != null) {
      params['keys'] = '';
    }
    params['separator'] = separator;
    var uri = this._buildUri(key, params);

    return (await this._client.get(uri, headers: headers));
  }

  /// Creates/Updates key from KV.
  ///
  /// Parameter [key] and [value] are the only required parameters.
  /// Other parameters are optional.
  /// You can read when to use parameters here:
  /// [create/update key](https://www.consul.io/api/kv.html#create-update-key).
  ///
  ///     var consul = new ConsulKV();
  ///     var newKey = await consul.put('foo', 'bar');
  ///     var insertIfNotExists = await consul.put('foo', 'bar', cas: 1);
  ///     print(newKey.body); // true
  ///     print(newKey.statusCode); // 200
  ///     print(insertIfNotExists.body); // false
  ///     print(insertIfNotExists.statusCode); // 200
  Future<http.Response> put(String key, String value,
      {String dc,
      int flags,
      int cas,
      String acquire,
      String release,
      Map<String, String> headers}) async {
    var params = <String, dynamic>{};
    headers = this._setHeaders(headers);
    if (dc != null) {
      params['dc'] = dc;
    }
    if (flags != null) {
      params['flags'] = flags.toString();
    }
    if (cas != null) {
      params['cas'] = cas.toString();
    }
    if (acquire != null) {
      params['acquire'] = acquire;
    }
    if (release != null) {
      params['release'] = release;
    }

    var uri = this._buildUri(key, params);

    return (await this._client.put(uri, body: value, headers: headers));
  }

  /// Build URI for request.
  ///
  /// Creates endpoint for request and passes query parameters for [Uri].
  Uri _buildUri(String resource, Map<String, dynamic> params) {
    var uri = new Uri(
        scheme: this.scheme,
        host: this.host,
        port: this.port,
        path: this._apiVersion + this._path + resource,
        queryParameters: params);
    return uri;
  }

  /// Sets header for request.
  ///
  /// Checks if they are present in parameters and combines them with
  /// [defaultHeaders] (if they are present).
  Map<String, String> _setHeaders(Map<String, String> headers) {
    if (headers == null) {
      headers = {};
    }
    if (this.defaultHeaders == null) {
      return headers;
    }
    headers.addAll(this.defaultHeaders);
    return headers;
  }
}
